crates/handlers/src/oauth2/registration.rs

```diff
@@ -219,8 +310,28 @@ mod tests {
         registration::ClientRegistrationResponse,
     };
     use sqlx::PgPool;
+    use url::Url;
+
+    use crate::{
+        oauth2::registration::host_is_public_suffix,
+        test_utils::{init_tracing, RequestBuilderExt, ResponseExt, TestState},
+    };
 
-    use crate::test_utils::{init_tracing, RequestBuilderExt, ResponseExt, TestState};
+    #[test]
+    fn test_public_suffix_list() {
+        fn url_is_public_suffix(url: &str) -> bool {
+            host_is_public_suffix(&Url::parse(url).unwrap())
+        }
+
+        assert!(url_is_public_suffix("https://.com."));
+        assert!(url_is_public_suffix("https://co.uk"));
+        assert!(url_is_public_suffix("https://github.io"));
+        assert!(!url_is_public_suffix("https://example.com"));
+        assert!(!url_is_public_suffix("https://matrix-org.github.io"));
+        assert!(!url_is_public_suffix("http://localhost"));
+        assert!(!url_is_public_suffix("org.matrix:/callback"));
+        assert!(!url_is_public_suffix("http://somerandominternaldomain"));
```

Comment:
That's one relevant test to understand what's considered to be a public suffix and what is not

Reply:
I'm not convinced the `host.len()` check is tested anywhere?

Reply:
What do you mean? The empty host is tested by `org.matrix:/callback`, and that length test is there to cover the `http://.com/` case: in that case, the prefix will be `com` but the host will be `.com`. If we were doing `host.len() != suffix.len()`, we would miss that case.
Effectively, that inequality is a shorter version `(host.len() != suffix.len()) && (host.len() != (suffix.len() + 1))`

Reply:
Ah, I misunderstood what that's checking. I can see how that's being hit now.

Reply:
I made the comment around that logic more obvious in https://github.com/matrix-org/matrix-authentication-service/pull/1785/commits/7141d8692245dd31de795fe3c843e5cae79f0525